### PR TITLE
merge DD_VERSION and DD_ENV with "dd.trace.global.tags"

### PR DESCRIPTION
**Pull request by lpriima**
This change introduces merging of explicitly defined span properties and properties defined in   "dd.trace.global.tags". 

I ) For every property  we have 3 levels of defining it:
1. creating tracer with configuration via API and with explicit Properties argument with specifying this exact property inside this argument ( https://github.com/DataDog/dd-trace-java/blob/master/dd-trace-api/src/main/java/datadog/trace/api/Config.java#L1384 )
1. java system property
1. linux environment variable
1. java property file
We take next one only if it's not defined the previous one. Eg. So we look into linux environment variables only 

II ) We have special container properties `tags` and `global.tags` which allow to aggregate other properties inside it.  
1. `dd.tags`(DD_TAGS) has precedence over `dd.trace.global.tags`. It will be merging of tags. 

III ) The result `service` tag in span is set up by
1. property(env variable) `dd.service` (DD_SERVICE) first.
1. if `dd.service` (DD_SERVICE)  is not set the result `service` tag is set by  `dd.service.name` (DD_SERVICE_NAME)
1. if neither of above is set, the result `service` tag is set by default integration name or just default name
1. you can't define `service` tag through tags

### CODE REVIEW

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java [hunk 1]
```diff
@@ -375,11 +375,26 @@ private String profilingProxyPasswordMasker() {
...
-    addPropToMapIfDefinedByEnvironment(tagsPreMap, VERSION);
-    tags = Collections.unmodifiableMap(tagsPreMap);
-    globalTags = getMapSettingFromEnvironment(GLOBAL_TAGS, null);
+    {
+      final Map<String, String> tagsPreMap =
+          new HashMap<>(getMapSettingFromEnvironment(TAGS, null));
+      if (tagsPreMap != null && !tagsPreMap.isEmpty()) {
+        // we only populate this tags if we use 'dd.tags'. If we don't use it: we populate this tags
+        // to
+        // 'dd.trace.global.tags' and globalTags field
```
Comment (tylerbenson) [inline dd-trace-api/src/main/java/datadog/trace/api/Config.java:384]: weird formatting...

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java [hunk 2]
```diff
@@ -375,11 +376,20 @@ private String profilingProxyPasswordMasker() {
...
+        // to 'dd.trace.global.tags' and globalTags field
+        tags = getMapWithPropertiesDefinedByEnvironment(tagsPreMap, ENV, VERSION);
+      } else {
+        tags = Collections.emptyMap();
+      }
+    }
+    globalTags =
+        getMapWithPropertiesDefinedByEnvironment(
+            getMapSettingFromEnvironment(GLOBAL_TAGS, null), ENV, VERSION);
+
```
Comment (tylerbenson) [inline dd-trace-api/src/main/java/datadog/trace/api/Config.java:392]: Do you think it would simplify things to reduce the distinction between tags and global tags and combine them into a single map as soon as possible?  (End result being just a single `tags` field.)

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy [hunk 3]
```diff
@@ -1142,22 +1144,27 @@ class ConfigTest extends DDSpecification {
...
+    prop.setProperty(Config.ENV, "eu-east")
+    prop.setProperty(Config.VERSION, "43")
 
     when:
     Config config = Config.get(prop)
 
     then:
-    config.mergedSpanTags == [a: "1", c: "3"]
-    config.mergedJmxTags == [a: "1", d: "4", (RUNTIME_ID_TAG): config.getRuntimeId(), (SERVICE_TAG): config.serviceName]
+    config.mergedSpanTags == [a: "1", c: "3",  (Config.ENV) : "eu-east", (Config.VERSION) : "43"]
```
Comment (tylerbenson) [inline dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1160]: When you've implemented merging of the tags/global tags, this should also have `b:2`, right?

File: dd-trace-api/src/main/java/datadog/trace/api/Config.java (hunk 2 above)
Comment (jdgumz) [reply dd-trace-api/src/main/java/datadog/trace/api/Config.java:392]: +1 I think this idea makes sense too. We just need to be careful about the order of precedence for updating the single map.

E.g.,
- Look for `DD_` prefixed env var first for env/service/version.
- Then look at JMX flags.

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy [hunk 4]
```diff
@@ -1276,11 +1283,191 @@ class ConfigTest extends DDSpecification {
...
+    def config = new Config()
+
+    then:
+    config.serviceName == DEFAULT_SERVICE_NAME
+    config.mergedSpanTags == [service:'service-tag-in-dd-trace-global-tags-java-property','service.version' : 'my-svc-vers']
+    config.mergedJmxTags == [(RUNTIME_ID_TAG): config.getRuntimeId(), (SERVICE_TAG): config.serviceName,
+                             'service.version' : 'my-svc-vers']
+  }
+
+  def "DD_SERVICE ignored when 'dd.service.name' java property is set; 'dd.service.name' overwrites DD_SERVICE_NAME"() {
```
Comment (jdgumz) [inline dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1403]: `DD_SERVICE` should not be ignored - it should have the highest precedence.

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy [hunk 5]
```diff
@@ -1276,11 +1283,191 @@ class ConfigTest extends DDSpecification {
...
+    environmentVariables.set("DD_SERVICE", "dd-service-env-var")
+    System.setProperty(PREFIX + GLOBAL_TAGS, "service:service-tag-in-dd-trace-global-tags-java-property,service.version:my-svc-vers")
+    environmentVariables.set(DD_GLOBAL_TAGS_ENV, "service:service-tag-in-env-var,service.version:my-svc-vers")
+
+    when:
+    def config = new Config()
+
+    then:
+    config.serviceName == "dd-service-env-var"
+    config.mergedSpanTags == [service:'service-tag-in-dd-trace-global-tags-java-property','service.version' : 'my-svc-vers']
```
Comment (jdgumz) [inline dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1465]: I'm not sure I understand. Why are we using the `service:service-tag-in-dd-trace-global-tags-java-property` still? I think it'd be confusing to see a different `service` in the span metadata from the one that should be used (the value of `DD_SERVICE`).

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 3 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1160]: if you want me to merge them. It might break existing backward compatibility because we already have several versions where `dd.tags` exclude `dd.trace.global.tags` is it ok ?

Comment (tylerbenson) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1160]: yeah, let's merge them into a single map inside `Config`.  Any edge cases where this would be problematic?

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 4 above)
Comment (tylerbenson) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1403]: In our codebase, we have a different order of precedence.  System Properties, defined on the command line have a higher precedence than environment variables.

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 5 above)
Comment (tylerbenson) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1465]: I assume `service` would get removed from the metadata before being reported to the agent, but I'm wondering if the part that does that would then replace the env-var service name since the rules for processing tags happen at a different place. This might be a problem and mean we need to pull those tags out earlier.

Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1465]: service name will be overwritten for span with the check above

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 3 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1160]: > yeah, let's merge them into a single map inside `Config`. Any edge cases where this would be problematic?

@jdgumz  ?

Comment (jdgumz) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1160]: The only thing I can think of is a hard-coded `service` tag existing in some Dockerfile or build config that goes unnoticed. If someone adds `DD_SERVICE` elsewhere and the value is different, it might be a surprise.

But I think as long as we call out the order of precedence very clearly in our release notes, folks should be able to triage the issue. I don't think we can support differing service values - there has to be one.

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 4 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1403]: @tylerbenson here @jdgumz wants java property(env var) `dd.service`(DD_SERVICE) take precedence over `dd.service.name`(DD_SERVICE_NAME).  That's another potentially backward incompatible change which might affect some deployments where we have both of them already. It looks like previous change where we introduced `tags` which overwrite old existing `global.tags` : https://github.com/DataDog/dd-trace-java/pull/1263/files#diff-43038f0f76f33bbf93ca377b0f77ff7aR707

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 5 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1465]: The point of this and some other test: we can have `service` tag in tags, which will be later ignored and FOR SURE will be overwritten by some other things up to default . That's behaviour is different from `env` and `version` tags, which will not be overwritten unless  corresponding java property or env var is set.

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy [hunk 6]
```diff
@@ -1276,11 +1282,190 @@ class ConfigTest extends DDSpecification {
...
+  }
+
+  // service name precedence checks
+  def "default service name exist"() {
     expect:
-    Config.propertyNameToEnvironmentVariableName(Config.SERVICE) == "DD_SERVICE"
+    Config.get().serviceName == DEFAULT_SERVICE_NAME
+  }
+
+  def "default service name is not affected by tags, nor env variables"() {
```
Comment (jdgumz) [inline dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1388]: I'm not sure I understand this case. If we have a `service` somewhere (global tags, or even `DD_SERVICE`) shouldn't that be used?

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 5 above)
Comment (jdgumz) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1463]: Gotcha. One other question: it seems like `DD_SERVICE` overrides the value of service even when there's a service tag in the global system property. This seems inconsistent with `dd.service` as a system property overriding the env var `DD_SERVICE`? It sounded like we wanted to maintain the highest precedence for system properties?

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy [hunk 7]
```diff
@@ -1276,11 +1282,190 @@ class ConfigTest extends DDSpecification {
...
+    def config = new Config()
+
+    then:
+    config.serviceName == DEFAULT_SERVICE_NAME
+    config.mergedSpanTags == [service:'service-tag-in-dd-trace-global-tags-java-property','service.version' : 'my-svc-vers']
+    config.mergedJmxTags == [(RUNTIME_ID_TAG): config.getRuntimeId(), (SERVICE_TAG): config.serviceName,
+                             'service.version' : 'my-svc-vers']
   }
 
+  def "DD_SERVICE precedence over 'dd.service.name' java property is set; 'dd.service' overwrites DD_SERVICE"() {
```
Comment (jdgumz) [inline dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1402]: Wait - don't we have it set up to allow `DD_ENV` and `DD_VERSION` to override `dd.trace.global.tags`?

I'm guessing this is a different case because `dd.service` already existed and generally system properties take the highest precedent?

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 5 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1463]: we have precedence of 
1. java system properties  
1. environment variables
1. property file

before, and implications of this might be more unpredictable for many other deployment. Of we want to change this precedence order, let's address it in a separate PR

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 6 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1388]: It should not. It will be ignored. That's requirement we have now. You can't define service name through tags. If you try to do so: it will be overwritten later

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 7 above)
Comment (lpriima) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1402]: > Wait - don't we have it set up to allow `DD_ENV` and `DD_VERSION` to override `dd.trace.global.tags`?

we don't have to, but if we set these 2 env vars they override corresponding tags. Here is a test for it:
https://github.com/DataDog/dd-trace-java/pull/1393/files#diff-19302d05d6472f1ef9c915823a88f122R1323

> I'm guessing this is a different case because `dd.service` already existed and generally system properties take the highest precedent?

File: dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy (hunk 6 above)
Comment (tylerbenson) [reply dd-trace-api/src/test/groovy/datadog/trace/api/ConfigTest.groovy:1388]: While this is true from the config perspective, we may have a problem with this when it is actually applied to a span.  Spans have decorators that are applied to tags.  Specifically `ServiceNameDecorator` may be problematic.  I think we want to update `SpanDecoratorTest` to validate this.
